lib/client/weblogin.go

```diff
@@ -364,6 +385,78 @@ func SSHAgentLogin(ctx context.Context, login SSHLoginDirect) (*auth.SSHLoginRes
 	return out, nil
 }
 
+// SSHAgentPwdlessLogin requests a passwordless MFA challenge via the proxy.
+// If the credentials are valid, the proxy will return a challenge. We then
+// prompt the user to touch a device and if user wasn't explicitly defined,
+// also prompts the user to select a login name. If the authentication succeeds,
+// we will get a temporary certificate back.
+func SSHAgentPwdlessLogin(ctx context.Context, login SSHLoginPasswordless) (*auth.SSHLoginResponse, error) {
+	webClient, webURL, err := initClient(login.ProxyAddr, login.Insecure, login.Pool)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	challengeJSON, err := webClient.PostJSON(
+		ctx, webClient.Endpoint("webapi", "mfa", "login", "begin"),
+		&MFAChallengeRequest{
+			Passwordless: true,
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	challenge := &MFAAuthenticateChallenge{}
+	if err := json.Unmarshal(challengeJSON.Bytes(), challenge); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// Sanity check WebAuthn challenge.
+	switch {
+	case challenge.WebauthnChallenge == nil:
+		return nil, trace.BadParameter("passwordless: webauthn challenge missing")
+	case challenge.WebauthnChallenge.Response.UserVerification == protocol.VerificationDiscouraged:
+		return nil, trace.BadParameter("passwordless: user verification requirement too lax (%v)", challenge.WebauthnChallenge.Response.UserVerification)
+	}
+
+	// Only pass on the user if explicitly set, otherwise let the credential
+	// picker kick in.
+	user := ""
+	if login.ExplicitUsername {
```

Comment:
Push this out? The caller can choose whether to set the User field or not.

Reply:
i might not be understanding, but the login is type `SSHLoginPasswordless` that allows you to set this flag from outside? eg:

```
response, err := SSHAgentPwdlessLogin(ctx, SSHLoginPasswordless{
		SSHLogin: SSHLogin{
			ProxyAddr:         tc.WebProxyAddr,
			PubKey:            pubKey,
			TTL:               tc.KeyTTL,
			Insecure:          tc.InsecureSkipVerify,
			Pool:              loopbackPool(tc.WebProxyAddr),
			Compatibility:     tc.CertificateFormat,
			RouteToCluster:    tc.SiteName,
			KubernetesCluster: tc.KubernetesCluster,
		},
		User:                    tc.Username,
		AuthenticatorAttachment: tc.AuthenticatorAttachment,
		ExplicitUsername:        tc.ExplicitUsername,   // <---- here
		Stderr:                  tc.Stderr,
	})
```

Reply:
We don't need the additional flag, the caller can just set the User field accordingly. In the previous code both User and ExplicitUsername are part of the TeleportClient, but here we don't need the distinction.

Reply:
oh right! :woman_facepalming: